### DISCUSSION TITLE
How to secure only SPECIFIC directories and/or files with JWT?

### DISCUSSION THREAD

**Discussion by nkosi23** [seaweedfs/seaweedfs#2926] [Q&A]
Looking at the wiki It looks like JWT can only be enabled globally (ie. for a whole volume server or for a whole Filer). Is there a way to enable JWT only for specific directories? 

My use case is that in a web application, there are both user uploaded images that must be fetched by the browser (for example as part of the content of a news article) and there are more sensitive content items such as invoices and other private documents for which I needs to authorize reads using JWT. 

Having JWT authorization for every image to be loaded on a page would be wasteful (generating a JWT token for every user uploaded asset that needs to be read on a page may add non-negligible and unnecessarily load). Similarly, having 2 filers (one with JWT enabled and one without) sounds overkill to me. Regarding this last option, I feel like this complexity will bite me in the back. For (lack of a better) example if I want to secure a directory but later change my mind, I would now need to migrate the files to the other filer (more maintenance). I do not know if such a situation is realistic but my guts make me feel uncomfortable with the idea of adding complexity by managing 2 filers. Managing this sort of business logic concern using the architecture of the storage layer feels wrong, I feel strongly that this will bite me down the line, but  who knows maybe this is the way to go.... 

What are you thoughts, is there a better way to achieve what I am after? I do not want to rely on third party systems such as Nginx/apache to manage this, I want everything to use solutions being native to Seaweedfs to keep the number of moving parts of the system low and manage everything in a single place rather than spreading authorization logic everywhere.

**Comment by nkosi23** [comment]
I am surprised by the silence as this looks like a common use case, maybe people implement such access control policies using Seaweedfs-based S3 Buckets? I'm curious to know if there is a native way to achieve this behavior since the S3 implementation comes with overheads and limitations (number of volumes used for every collections etc...) while using "bare"-Seaweedfs and JWT has many benefits (in addition to the fact that we do not use S3 anywhere so it looks overkill to introduce the whole S3 API in our storage layer just for access control).

**Comment by chrislusf** [comment]
If this MUST be implemented, the most efficient way is to use path-specific configuration to specify the JWT tokens. The configuration is kind of cumbersome. In stead of one global setting, you can configure jwt signing key for specific locations.

**Reply by nkosi23** [reply]
Thanks a lot for dropping by, really appreciate it, could you please give me pointers on where I could learn about using path-specific configuration to specify the JWT tokens / configuring jwt for specific locations? If no wiki page exists even just pointing me to the right source code file would be massively helpful!

**Reply by chrislusf** [reply]
this "path-specific configuration to specify the JWT tokens" feature does not exist. The above answer is just one possible design. The related wiki is https://github.com/chrislusf/seaweedfs/wiki/Path-Specific-Configuration
